### PR TITLE
feat(chisel): display return value on function call

### LINKED ISSUE #5481: `chisel` doesn't display the return value from a function call unless assigned to a variable
### Component

Chisel

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (6e09b4e 2023-07-24T00:24:06.799680000Z)

### What command(s) is the bug in?

chisel

### Operating System

macOS (Apple Silicon)

### Describe the bug

`chisel` doesn't display the return value from a function call unless assigned to a variable.

Example:

```
➜ MerkleProof.processProof(proof, leaf)
➜ bytes32 output = MerkleProof.processProof(proof, leaf)
➜ output
Type: bytes32
└ Data: 0xe90b7bceb6e7df5418fb78d8ee546e97c83a08bbccc01a0644d599ccd2a7c2e0
```

It'd be nice if the first `➜` had the same output as the third `➜`.

If that's not possible for some technical reason, it'd be helpful to document this behavior in the docs. (I see an example with an assignment like that [here](https://github.com/foundry-rs/foundry/tree/master/chisel#toggling-traces) but it's not clear that assignments are mandatory!)

PS: I also tried `console.logBytes32(MerkleProof.processProof(proof, leaf));` but that also displayed nothing (unless `!traces` was enabled).

### PR DESCRIPTION
**Pull request by teddav**
Closes #5481 

## Solution
Check if the current action is a function call. If so, display the result of the call

### CODE REVIEW

File: chisel/src/executor.rs [hunk 1]
```diff
@@ -988,6 +997,42 @@ impl Type {
...
+        let Some(pt::Expression::Variable(contract_name)) =
+            intermediate.repl_contract_expressions.get(&contract_name.name)
+        else {
+            return None
+        };
+
+        let Some(contract) = intermediate
+            .intermediate_contracts
+            .get(&contract_name.name)
+            .unwrap()
```
Comment (mattsse) [inline chisel/src/executor.rs:1024]:
```suggestion
            ?
```

File: chisel/src/executor.rs [hunk 2]
```diff
@@ -988,6 +997,42 @@ impl Type {
...
+            .function_definitions
+            .get(&function_name.name)
+        else {
+            return None
+        };
+
+        let Some(returns) = contract.as_ref().returns.first() else { return None };
```
Comment (mattsse) [inline chisel/src/executor.rs:1031]: can use `?` operator here

File: chisel/src/executor.rs [hunk 3]
```diff
@@ -988,6 +997,42 @@ impl Type {
...
+        let Some(return_parameter) = returns.to_owned().1 else { return None };
```
Comment (mattsse) [inline chisel/src/executor.rs:1032]: same

File: chisel/src/executor.rs [hunk 4]
```diff
@@ -176,12 +176,21 @@ impl SessionSource {
...
-            Some(res) => res,
-            // this type was denied for inspection, continue
-            None => return Ok((true, None)),
+        // If the current action is a function call, we get its return type
+        // otherwise it returns None
+        let function_call_return_type =
+            Type::get_function_return_type(contract_expr, &generated_output.intermediate);
+
+        let (contract_expr, ty) = if function_call_return_type.is_some() {
+            (contract_expr.unwrap(), function_call_return_type.unwrap())
```
Comment (mattsse) [inline chisel/src/executor.rs:185]: I think this can be an `if let (Some(), Some())` so we can get rid of the unwraps here

**Review by mattsse — CHANGES_REQUESTED** [review]
cool!

changes look good,
some style nits

**Comment by teddav** [comment]
thanks for the review @mattsse 
should be a bit cleaner now

File: chisel/src/executor.rs [hunk 5]
```diff
@@ -176,12 +176,22 @@ impl SessionSource {
...
+        // otherwise it returns None
+        let function_call_return_type =
+            Type::get_function_return_type(contract_expr, &generated_output.intermediate);
+
+        let (contract_expr, ty) = if let Some(function_call_return_type) = function_call_return_type
         {
-            Some(res) => res,
-            // this type was denied for inspection, continue
-            None => return Ok((true, None)),
+            (contract_expr.unwrap(), function_call_return_type)
```
Comment (Evalir) [inline chisel/src/executor.rs:186]: Can we also avoid unwrapping here?

File: chisel/src/executor.rs [hunk 6]
```diff
@@ -988,6 +998,31 @@ impl Type {
...
+        let pt::Expression::MemberAccess(_, contract_name, function_name) = function_call.as_ref()
+        else {
+            return None
+        };
+        let pt::Expression::Variable(contract_name) = contract_name.as_ref() else { return None };
+        let pt::Expression::Variable(contract_name) =
+            intermediate.repl_contract_expressions.get(&contract_name.name)?
+        else {
+            return None
+        };
```
Comment (Evalir) [inline chisel/src/executor.rs:1015]: I don't understand what's happening here 😅 can't we rewrite this so it's a little clearer and add docs?

Comment (teddav) [reply chisel/src/executor.rs:1015]: `contract_expr` represents the input of the user as a tree Expression. Im just trying to unwrap it and get to the string itself. I couldnt find an easier way to get to it except unwrapping each level 1 by 1.
If you have a better way, I'll change it.
I can add some comments if needed

**Comment by Evalir** [comment]
hey @teddav pushed a few commits to just tidy up a bit, hope you don't mind!

let us know when this is working and ping us if you need help 🙏

**Comment by teddav** [comment]
I don't see how the match makes it clearer, but if you prefer, I don't mind. I think the let..else was more concise.
If it's good for you, I guess it can be merged then

**Review by Evalir — APPROVED** [review]

**Pull request merged**